File: tarka/lexer.py
"""
Lexer for Tarka Language
"""

from tarka.errors import LexerError
from tarka.token import GetKeywordOrIdentifierTokenType, Token, TokenType


class Lexer:
    """
    Lexer class
    """

    def __init__(self, code: str) -> None:
        """
        Creates instance of Lexer class

        :param self: Lexer
        :param code: Code to tokenize
        :type code: str
        """
        self.code: str = code
        self.codeLen: int = len(code)
        self.curPos = int(-1)
        self.curChar = str("\0")
        self.line = int(1)
        self.col = int(0)

    def GetTokens(self) -> list[Token]:
        """
        Returns list of tokens tokenized from Token

        :param self: Lexer
        :return: List of tokens
        :rtype: list[Token]
        """
        tokens: list[Token] = []
        self.__readChar()
        while self.curPos < self.codeLen:
            if self.curChar in [" ", "\r", "\t"]:
                self.__readWhiteSpace()
            elif self.curChar == "\n":
                tok = self.__readNewline()
                tokens.append(tok)
            elif self.curChar.isalpha() or self.curChar == "_":
                tok = self.__readIdentifier()
                tokens.append(tok)
            elif self.curChar.isdigit():
                tok = self.__readNumber()
                tokens.append(tok)
            else:
                raise LexerError(self.curChar, self.line, self.col)
        return tokens

    def __readChar(self) -> None:
        """
        Moves the current pointer to the next character

        :param self: Lexer
        """
        self.curPos += 1
        if self.curPos >= self.codeLen:
            return
        if self.curChar == "\n":
            self.line += 1
            self.col = 0
        self.curChar = self.code[self.curPos]
        self.col += 1

    def __readWhiteSpace(self) -> None:
        """
        Skips the whitespace

        :param self: Lexer
        """
        self.__readChar()

    def __readNewline(self) -> Token:
        """
        Reads the Newline

        :param self: Lexer
        """
        tok = Token(TokenType.EX_NEWLINE, "\n", self.line, self.col)
        self.__readChar()
        return tok

    def __readIdentifier(self) -> Token:
        """
        Returns the Identifier or Keyword token

        :param self: Lexer
        :return: Identifier or Keyword token
        :rtype: Token
        """
        beg: int = self.curPos
        line: int = self.line
        col: int = self.col
        while self.curPos < self.codeLen and (self.curChar.isalpha() or self.curChar == "_"):
            self.__readChar()
        ident = self.code[beg : self.curPos]
        tokType = GetKeywordOrIdentifierTokenType(ident)
        return Token(tokType, ident, line, col)

    def __readNumber(self) -> Token:
        """
        Returns the specific number token

        :param self: Lexer
        :return: Number Token
        :rtype: Token
        """
        beg: int = self.curPos
        line: int = self.line
        col: int = self.col
        while self.curPos < self.codeLen and self.curChar.isdigit():
            self.__readChar()
        number = self.code[beg : self.curPos]
        return Token(TokenType.LIT_INT, number, line, col)

File: lexer/lexer.go
package lexer

import (
	"github.com/adinovap20/tarka/token"
)

type Lexer struct {
	code    string
	pos     int
	readPos int
	ch      byte
}

func New(code string) *Lexer {
	l := &Lexer{code: code}
	l.readChar()
	return l
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	l.skipWhitespace()

	switch l.ch {
	case '\n':
		tok.Lit = "\n"
		tok.Type = token.EX_NEWLINE
	case 0:
		tok.Lit = ""
		tok.Type = token.EX_EOF
	default:
		if isLetter(l.ch) {
			tok.Lit = l.readIdentifier()
			tok.Type = token.LookupIdent(tok.Lit)
			return tok
		} else if isDigit(l.ch) {
			tok.Lit = l.readNumber()
			tok.Type = token.LIT_INT
			return tok
		} else {
			tok.Type = token.EX_UNKNOWN
			tok.Lit = ""
		}
	}

	l.readChar()
	return tok
}

func (l *Lexer) readChar() {
	if l.readPos >= len(l.code) {
		l.ch = 0
	} else {
		l.ch = l.code[l.readPos]
	}
	l.pos = l.readPos
	l.readPos++
}

func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\r' {
		l.readChar()
	}
}

func (l *Lexer) readIdentifier() string {
	position := l.pos
	for isAlnum(l.ch) {
		l.readChar()
	}
	return l.code[position:l.pos]
}

func (l *Lexer) readNumber() string {
	position := l.pos
	for isDigit(l.ch) {
		l.readChar()
	}
	return l.code[position:l.pos]
}

func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func isAlnum(ch byte) bool {
	return isLetter(ch) || isDigit(ch)
}
